File: WdtcCore/src/main/java/org/wdt/wdtc/game/config/GameConfig.java
package org.wdt.wdtc.game.config;

import lombok.SneakyThrows;
import org.apache.log4j.Logger;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.game.DownloadedGameVersion;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.manger.GameFolderManger;
import org.wdt.wdtc.utils.WdtcLogger;
import org.wdt.wdtc.utils.gson.JSON;
import org.wdt.wdtc.utils.gson.JSONUtils;

import java.io.IOException;
import java.util.List;

public class GameConfig {
    private static final Logger logmaker = WdtcLogger.getLogger(WdtcLogger.class);
    private final Launcher launcher;

    public GameConfig(Launcher launcher) {
        this.launcher = launcher;
    }

    public static void writeConfigJsonToAllVersion() throws IOException {
        GameFolderManger gameFolderManger = new GameFolderManger();
        if (DownloadedGameVersion.isDownloadedGame(gameFolderManger)) {
            List<Launcher> list = DownloadedGameVersion.getGameVersionList(gameFolderManger);
            for (Launcher child : list) {
                GameConfig config = child.getGameConfig();
                if (FileUtils.isFileNotExists(child.getVersionConfigFile())) {
                    writeConfigJson(child);
                } else {
                    DefaultGameConfig GameConfig = config.getDefaultGameConfig();
                    GameConfig.setInfo(config.getVersionInfo());
                    logmaker.info(GameConfig);
                    config.putConfigToFile(GameConfig);
                }
            }
        } else {
            logmaker.warn("There are currently no game versions available");
        }
    }


    public static void writeConfigJson(Launcher launcher) {
        try {
            DefaultGameConfig config = new DefaultGameConfig(launcher);
            FileUtils.writeStringToFile(launcher.getVersionConfigFile(), JSON.GSONBUILDER.serializeNulls().setPrettyPrinting().create().toJson(config));
            logmaker.info(launcher.getVersionNumber() + " " + config);
        } catch (IOException e) {
            logmaker.error(WdtcLogger.getErrorMessage(e));
        }
    }

    public DefaultGameConfig.Config getConfig() {
        return getDefaultGameConfig().getConfig();
    }

    public static void CkeckVersionInfo(Launcher launcher) throws IOException {
        DefaultGameConfig config = launcher.getGameConfig().getDefaultGameConfig();
        config.setInfo(launcher.getVersionInfo());
        FileUtils.writeStringToFile(launcher.getVersionConfigFile(), JSON.GSONBUILDER.serializeNulls().setPrettyPrinting().create().toJson(config));
        logmaker.info(launcher.getVersionNumber() + " " + config);
    }

    @SneakyThrows(IOException.class)
    public DefaultGameConfig getDefaultGameConfig() {
        return JSONUtils.readJsonFileToClass(launcher.getVersionConfigFile(), DefaultGameConfig.class);
    }

    public VersionInfo getVersionInfo() {
        return getDefaultGameConfig().getInfo();

    }

    public void putConfigToFile(DefaultGameConfig config) throws IOException {
        FileUtils.writeStringToFile(launcher.getVersionConfigFile(),
                JSON.GSONBUILDER.serializeNulls().setPrettyPrinting().create().toJson(config));
    }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/game/config/GameConfig.kt
package org.wdt.wdtc.core.game.config

import org.wdt.utils.gson.Json
import org.wdt.utils.gson.readFileToClass
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.utils.io.isFileNotExists
import org.wdt.wdtc.core.game.*
import org.wdt.wdtc.core.manger.GameDirectoryManger
import org.wdt.wdtc.core.utils.getExceptionMessage
import org.wdt.wdtc.core.utils.logmaker
import java.io.IOException

class GameConfig(private val launcher: Launcher) {
  val config: DefaultGameConfig.Config?
    get() = defaultGameConfig.config

  val defaultGameConfig: DefaultGameConfig
    get() = launcher.versionConfigFile.readFileToClass()
  val versionInfo: VersionInfo?
    get() = defaultGameConfig.info

  fun putConfigToFile(config: DefaultGameConfig) {
    launcher.versionConfigFile.writeObjectToFile(
      config, Json.getBuilder().setPrettyPrinting()
    )
  }
}

fun writeConfigJsonToAllVersion() {
  val gameDirectoryManger = GameDirectoryManger()
  if (gameDirectoryManger.isDownloadedGame) {
    gameDirectoryManger.getGameVersionList()?.forEach {
      val config = it.gameConfig
      if (it.versionConfigFile.isFileNotExists()) {
        it.writeConfigJson()
      } else {
        val gameConfig = config.defaultGameConfig
        gameConfig.info = config.versionInfo
        logmaker.info(gameConfig)
        config.putConfigToFile(gameConfig)
      }
    }
  } else {
    logmaker.warn("There are currently no game versions available")
  }
}

fun Launcher.writeConfigJson() {
  try {
    val config = DefaultGameConfig(this)
    this.versionConfigFile.writeObjectToFile(
      config, Json.getBuilder().setPrettyPrinting()
    )
    logmaker.info("${this.versionNumber} $config")
  } catch (e: IOException) {
    logmaker.error(e.getExceptionMessage())
  }
}

@Throws(IOException::class)
fun Launcher.ckeckVersionInfo() {
  val config = this.gameConfig.defaultGameConfig
  config.info = this.versionInfo
  this.versionConfigFile.writeObjectToFile(
    config, Json.getBuilder().setPrettyPrinting()
  )
  logmaker.info("${this.versionNumber} $config")
}

val Launcher.gameConfig: GameConfig
  get() = GameConfig(this)
